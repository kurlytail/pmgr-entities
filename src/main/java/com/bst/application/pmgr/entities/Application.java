package com.bst.application.pmgr.entities;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.boot.web.servlet.support.SpringBootServletInitializer;
import org.springframework.context.annotation.ComponentScan;
import org.springframework.transaction.annotation.EnableTransactionManagement;

import com.bst.configuration.pmgr.entities.PmgrEntitiesConfiguration;

@SpringBootApplication
@ComponentScan(basePackageClasses = { PmgrEntitiesConfiguration.class, Application.class })
@EnableTransactionManagement
public class Application extends SpringBootServletInitializer {

	public static void main(final String[] args) {
		SpringApplication.run(Application.class, args);
	}
}