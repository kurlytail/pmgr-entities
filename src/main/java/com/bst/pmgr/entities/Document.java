package com.bst.pmgr.entities;

import java.util.List;

import javax.persistence.DiscriminatorColumn;
import javax.persistence.Entity;
import javax.persistence.EntityListeners;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.Inheritance;
import javax.persistence.InheritanceType;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;

import com.bst.pmgr.entities.audit.DocumentAudit;

import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.Setter;
import lombok.ToString;

@EntityListeners(DocumentAudit.class)
@Entity
@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
@DiscriminatorColumn(name = "documentType")
@ToString
@Getter
@EqualsAndHashCode
public class Document {
	@Id
	@GeneratedValue(strategy = GenerationType.AUTO)
	private Long id;

	@Setter
	private String name;

	@ManyToOne
	@Setter
	private Work work;

	@OneToMany(mappedBy = "document")
	private List<Section> sections;
}
