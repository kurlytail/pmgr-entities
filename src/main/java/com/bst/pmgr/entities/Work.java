package com.bst.pmgr.entities;

import java.util.ArrayList;
import java.util.List;

import javax.persistence.Column;
import javax.persistence.DiscriminatorColumn;
import javax.persistence.Entity;
import javax.persistence.EntityListeners;
import javax.persistence.FetchType;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.Inheritance;
import javax.persistence.InheritanceType;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;
import javax.persistence.Table;

import com.bst.pmgr.entities.audit.WorkAudit;
import com.fasterxml.jackson.annotation.JsonIdentityInfo;
import com.fasterxml.jackson.annotation.ObjectIdGenerators;

@EntityListeners(WorkAudit.class)
@Entity
@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
@DiscriminatorColumn(name = "workType")
@JsonIdentityInfo(generator = ObjectIdGenerators.PropertyGenerator.class, property = "id")
@Table(name = "WORK_TABLE")
public class Work {	
	@Id
	@GeneratedValue(strategy = GenerationType.AUTO)
	private Long id;

	@Column(nullable = false)
	private String name;

	@OneToMany(mappedBy = "work")
	private final List<Activity> activities = new ArrayList<>();

	@OneToMany(mappedBy = "work")
	private final List<Document> documents = new ArrayList<>();

	@OneToMany(mappedBy = "work")
	private final List<Tool> tools = new ArrayList<>();

	@OneToMany(mappedBy = "work")
	private final List<Process> processes = new ArrayList<>();

	@OneToMany(mappedBy = "work")
	private final List<ProcessGroup> processGroups = new ArrayList<>();

	@ManyToOne(fetch = FetchType.LAZY)
	private PmgrPerson createdBy;

	@Override
	public boolean equals(final Object obj) {
		if (this == obj) {
			return true;
		}
		if (obj == null) {
			return false;
		}
		if (this.getClass() != obj.getClass()) {
			return false;
		}
		final Work other = (Work) obj;
		if (this.activities == null) {
			if (other.activities != null) {
				return false;
			}
		} else if (!this.activities.equals(other.activities)) {
			return false;
		}
		if (this.createdBy == null) {
			if (other.createdBy != null) {
				return false;
			}
		} else if (!this.createdBy.equals(other.createdBy)) {
			return false;
		}
		if (this.documents == null) {
			if (other.documents != null) {
				return false;
			}
		} else if (!this.documents.equals(other.documents)) {
			return false;
		}
		if (this.id == null) {
			if (other.id != null) {
				return false;
			}
		} else if (!this.id.equals(other.id)) {
			return false;
		}
		if (this.name == null) {
			if (other.name != null) {
				return false;
			}
		} else if (!this.name.equals(other.name)) {
			return false;
		}
		if (this.processGroups == null) {
			if (other.processGroups != null) {
				return false;
			}
		} else if (!this.processGroups.equals(other.processGroups)) {
			return false;
		}
		if (this.processes == null) {
			if (other.processes != null) {
				return false;
			}
		} else if (!this.processes.equals(other.processes)) {
			return false;
		}
		if (this.tools == null) {
			if (other.tools != null) {
				return false;
			}
		} else if (!this.tools.equals(other.tools)) {
			return false;
		}
		return true;
	}

	public List<Activity> getActivities() {
		return this.activities;
	}

	public PmgrPerson getCreatedBy() {
		return this.createdBy;
	}

	public List<Document> getDocuments() {
		return this.documents;
	}

	public Long getId() {
		return this.id;
	}

	public String getName() {
		return this.name;
	}

	public List<Process> getProcesses() {
		return this.processes;
	}

	public List<ProcessGroup> getProcessGroups() {
		return this.processGroups;
	}

	public List<Tool> getTools() {
		return this.tools;
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + ((this.activities == null) ? 0 : this.activities.hashCode());
		result = prime * result + ((this.createdBy == null) ? 0 : this.createdBy.hashCode());
		result = prime * result + ((this.documents == null) ? 0 : this.documents.hashCode());
		result = prime * result + ((this.id == null) ? 0 : this.id.hashCode());
		result = prime * result + ((this.name == null) ? 0 : this.name.hashCode());
		result = prime * result + ((this.processGroups == null) ? 0 : this.processGroups.hashCode());
		result = prime * result + ((this.processes == null) ? 0 : this.processes.hashCode());
		result = prime * result + ((this.tools == null) ? 0 : this.tools.hashCode());
		return result;
	}

	public void setCreatedBy(final PmgrPerson createdBy) {
		this.createdBy = createdBy;
	}

	public void setName(final String name) {
		this.name = name;
	}

	@Override
	public String toString() {
		return "Work [id=" + this.id + ", name=" + this.name + ", createdBy=" + this.createdBy + "]";
	}

}
