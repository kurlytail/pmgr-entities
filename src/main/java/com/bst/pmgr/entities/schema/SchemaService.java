package com.bst.pmgr.entities.schema;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

import javax.annotation.PostConstruct;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.io.ResourceLoader;
import org.springframework.stereotype.Service;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;

@Service
public class SchemaService {

	static final Logger LOG = LoggerFactory.getLogger(SchemaService.class);

	@Autowired
	private ResourceLoader resourceLoader;

	private Map<String, SchemaActivity> schemaMap = new HashMap<>();
	private Map<String, MetaActivity> activities = new HashMap<>();
	private Map<String, MetaDocument> documents = new HashMap<>();
	private Map<String, MetaTool> tools = new HashMap<>();
	private Map<String, MetaProcess> processes = new HashMap<>();

	private Map<String, MetaProcessGroup> processGroups = new HashMap<>();
	
	@PostConstruct
	public void init() throws JsonProcessingException, IOException {
		final ObjectMapper objectMapper = new ObjectMapper();
		this.schemaMap = objectMapper.readValue(resourceLoader.getResource("classpath:schema.json").getInputStream(),
				new TypeReference<Map<String, SchemaActivity>>() {
				});

		this.createSkeletonMaps();
		this.fixupActivities();
		this.fixupDocuments();
		this.fixupTools();
		this.fixupProcesses();
		this.fixupProcessGroups();
	}

	private void createSkeletonMaps() {
		this.schemaMap.forEach((name, meta) -> {
			final MetaActivity metaActivity = new MetaActivity();
			metaActivity.setName(name);
			this.activities.put(name, metaActivity);

			meta.getInputs().forEach(inputName -> {
				final MetaDocument metaDocument = new MetaDocument();
				metaDocument.setName(inputName);
				this.documents.put(inputName, metaDocument);
			});

			meta.getOutputs().forEach(outputName -> {
				final MetaDocument metaDocument = new MetaDocument();
				metaDocument.setName(outputName);
				this.documents.put(outputName, metaDocument);
			});

			meta.getTools().forEach(toolName -> {
				final MetaTool metaTool = new MetaTool();
				metaTool.setName(toolName);
				this.tools.put(toolName, metaTool);
			});

			final MetaProcessGroup metaProcessGroup = new MetaProcessGroup();
			metaProcessGroup.setName(meta.getProcessGroup());
			this.processGroups.put(metaProcessGroup.getName(), metaProcessGroup);

			final MetaProcess metaProcess = new MetaProcess();
			metaProcess.setName(meta.getProcess());
			this.processes.put(metaProcess.getName(), metaProcess);
		});
	}

	private void fixupActivities() {
		this.activities.forEach((name, metaActivity) -> {
			final SchemaActivity schemaActivity = this.schemaMap.get(name);
			schemaActivity.getInputs().forEach(inputName -> {
				metaActivity.getInputs().put(inputName, this.documents.get(inputName));
			});
			schemaActivity.getOutputs().forEach(outputName -> {
				metaActivity.getOutputs().put(outputName, this.documents.get(outputName));
			});
			schemaActivity.getTools().forEach(toolName -> {
				metaActivity.getTools().put(toolName, this.tools.get(toolName));
			});
			metaActivity.setProcess(this.processes.get(schemaActivity.getProcess()));
			metaActivity.setProcessGroup(this.processGroups.get(schemaActivity.getProcessGroup()));
		});
	}

	private void fixupDocuments() {
		this.activities.forEach((name, metaActivity) -> {
			final SchemaActivity schemaActivity = this.schemaMap.get(name);

			schemaActivity.getInputs().forEach(inputName -> {
				final MetaDocument metaDocument = this.documents.get(inputName);
				metaDocument.getConsumingActivities().put(inputName, metaActivity);
				schemaActivity.getTools().forEach(toolName -> {
					metaDocument.getConsumingTools().put(toolName, this.tools.get(toolName));
				});
				metaDocument.getConsumingProcesses().put(schemaActivity.getProcess(),
						this.processes.get(schemaActivity.getProcess()));
				metaDocument.getConsumingProcessGroups().put(schemaActivity.getProcessGroup(),
						this.processGroups.get(schemaActivity.getProcessGroup()));
			});

			schemaActivity.getOutputs().forEach(outputName -> {
				final MetaDocument metaDocument = this.documents.get(outputName);
				metaDocument.getProducingActivities().put(outputName, metaActivity);
				schemaActivity.getTools().forEach(toolName -> {
					metaDocument.getProducingTools().put(toolName, this.tools.get(toolName));
				});
				metaDocument.getProducingProcesses().put(schemaActivity.getProcess(),
						this.processes.get(schemaActivity.getProcess()));
				metaDocument.getProducingProcessGroups().put(schemaActivity.getProcessGroup(),
						this.processGroups.get(schemaActivity.getProcessGroup()));
			});
		});
	}

	private void fixupProcesses() {
		this.activities.forEach((name, metaActivity) -> {
			final SchemaActivity schemaActivity = this.schemaMap.get(name);

			final String process = schemaActivity.getProcess();
			final MetaProcess metaProcess = this.processes.get(process);
			metaProcess.getActivities().put(name, metaActivity);
			schemaActivity.getInputs().forEach(inputName -> {
				metaProcess.getConsumedDocuments().put(inputName, this.documents.get(inputName));
			});
			schemaActivity.getOutputs().forEach(outputName -> {
				metaProcess.getProducedDocuments().put(outputName, this.documents.get(outputName));
			});
			metaProcess.setProcessGroup(this.processGroups.get(schemaActivity.getProcess()));
			schemaActivity.getTools().forEach(toolName -> {
				metaProcess.getTools().put(toolName, this.tools.get(toolName));
			});
		});
	}

	private void fixupProcessGroups() {
		this.activities.forEach((name, metaActivity) -> {
			final SchemaActivity schemaActivity = this.schemaMap.get(name);

			final String processGroupName = schemaActivity.getProcessGroup();
			final MetaProcessGroup metaProcessGroup = this.processGroups.get(processGroupName);
			metaProcessGroup.getActivities().put(name, metaActivity);
			schemaActivity.getInputs().forEach(inputName -> {
				metaProcessGroup.getConsumedDocuments().put(inputName, this.documents.get(inputName));
			});
			schemaActivity.getOutputs().forEach(outputName -> {
				metaProcessGroup.getProducedDocuments().put(outputName, this.documents.get(outputName));
			});
			metaProcessGroup.getProcesses().put(schemaActivity.getProcess(),
					this.processes.get(schemaActivity.getProcess()));
			schemaActivity.getTools().forEach(toolName -> {
				metaProcessGroup.getTools().put(toolName, this.tools.get(toolName));
			});
		});
	}

	private void fixupTools() {
		this.activities.forEach((name, metaActivity) -> {
			final SchemaActivity schemaActivity = this.schemaMap.get(name);

			schemaActivity.getTools().forEach(toolName -> {
				final MetaTool metaTool = this.tools.get(toolName);
				metaTool.getActivities().put(name, metaActivity);
				schemaActivity.getInputs().forEach(inputName -> {
					metaTool.getConsumedDocuments().put(inputName, this.documents.get(inputName));
				});
				metaActivity.getOutputs().forEach((outputName, metaDocument) -> {
					metaTool.getProducedDocuments().put(outputName, this.documents.get(outputName));
				});
				metaTool.getProcesses().put(schemaActivity.getProcess(),
						this.processes.get(schemaActivity.getProcess()));
				metaTool.getProcessGroups().put(schemaActivity.getProcessGroup(),
						this.processGroups.get(schemaActivity.getProcessGroup()));
			});
		});
	}

	public Map<String, MetaActivity> getActivities() {
		return this.activities;
	}

	public Map<String, MetaDocument> getDocuments() {
		return this.documents;
	}

	public Map<String, MetaProcess> getProcesses() {
		return this.processes;
	}

	public Map<String, MetaProcessGroup> getProcessGroups() {
		return this.processGroups;
	}

	public Map<String, SchemaActivity> getSchemaMap() {
		return this.schemaMap;
	}

	public Map<String, MetaTool> getTools() {
		return this.tools;
	}

	public void setActivities(final Map<String, MetaActivity> activities) {
		this.activities = activities;
	}

	public void setDocuments(final Map<String, MetaDocument> documents) {
		this.documents = documents;
	}

	public void setProcesses(final Map<String, MetaProcess> processes) {
		this.processes = processes;
	}

	public void setProcessGroups(final Map<String, MetaProcessGroup> processGroups) {
		this.processGroups = processGroups;
	}

	public void setSchemaMap(final Map<String, SchemaActivity> schemaMap) {
		this.schemaMap = schemaMap;
	}

	public void setTools(final Map<String, MetaTool> tools) {
		this.tools = tools;
	}

}
