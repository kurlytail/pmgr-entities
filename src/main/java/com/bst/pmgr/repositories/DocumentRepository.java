package com.bst.pmgr.repositories;

import org.springframework.data.repository.CrudRepository;

import com.bst.pmgr.entities.GenericDocument;

public interface DocumentRepository extends CrudRepository<GenericDocument, Long> {
}
