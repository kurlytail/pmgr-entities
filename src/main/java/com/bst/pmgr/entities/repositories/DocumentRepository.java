package com.bst.pmgr.entities.repositories;

import org.springframework.data.repository.CrudRepository;
import org.springframework.data.rest.core.annotation.RepositoryRestResource;

import com.bst.pmgr.entities.Document;

@RepositoryRestResource
public interface DocumentRepository extends CrudRepository<Document, Long> {
}
