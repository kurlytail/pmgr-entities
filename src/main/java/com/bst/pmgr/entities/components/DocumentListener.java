package com.bst.pmgr.entities.components;

import java.util.List;
import java.util.stream.Collectors;

import javax.persistence.PrePersist;

import org.kohsuke.randname.RandomNameGenerator;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import com.bst.pmgr.entities.Document;
import com.bst.pmgr.entities.Tool;
import com.bst.pmgr.entities.schema.SchemaService;

@Component
public class DocumentListener {

	private final RandomNameGenerator rnd = new RandomNameGenerator(0);

	@Autowired
	private SchemaService schemaService;

	@PrePersist
	public void prePersist(final Document document) throws Exception {
		if (document.getName() == null) {
			document.setName(this.rnd.next());
		}

		if (document.getMetaName() == null) {
			throw new Exception("Document does not have a meta name");
		}

		if (document.getTools().size() == 0) {
			final List<Tool> toolList = this.schemaService.getDocuments().get(document.getMetaName())
					.getProducingTools().values().stream().map(metaTool -> {
						final Tool tool = new Tool();
						tool.setMetaName(metaTool.getName());
						tool.setName(this.rnd.next());
						tool.setDocument(document);
						return tool;
					}).collect(Collectors.toList());

			document.getTools().addAll(toolList);
			document.getWork().getTools().addAll(toolList);
		}
	}

}
