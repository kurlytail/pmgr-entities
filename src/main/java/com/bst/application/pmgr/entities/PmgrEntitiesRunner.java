package com.bst.application.pmgr.entities;

import java.util.Random;

import javax.persistence.EntityManager;
import javax.transaction.Transactional;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.CommandLineRunner;
import org.springframework.stereotype.Service;

import com.bst.pmgr.entities.Work;
import com.bst.pmgr.entities.repositories.WorkRepository;

@Service
@Transactional
public class PmgrEntitiesRunner implements CommandLineRunner {

	@Autowired
	private WorkRepository workRepository;

	@Autowired
	private EntityManager entityManager;

	@Override
	public void run(final String... args) throws Exception {

		final long countOfWork = Math.abs(new Random().nextLong()) % 10;
		long indexOfWork = 0;
		
		while (indexOfWork < countOfWork) {
			final Work work = new Work();
			this.workRepository.save(work);
			++indexOfWork;
		}

		this.entityManager.flush();
	}

}
