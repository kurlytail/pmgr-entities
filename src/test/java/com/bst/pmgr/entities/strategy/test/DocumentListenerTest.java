package com.bst.pmgr.entities.strategy.test;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.autoconfigure.domain.EntityScan;
import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
import org.springframework.boot.test.context.TestConfiguration;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.EnableAspectJAutoProxy;
import org.springframework.data.jpa.repository.config.EnableJpaRepositories;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.TestExecutionListeners;
import org.springframework.test.context.TestExecutionListeners.MergeMode;
import org.springframework.test.context.TestPropertySource;
import org.springframework.test.context.junit.jupiter.SpringExtension;
import org.springframework.transaction.annotation.Propagation;
import org.springframework.transaction.annotation.Transactional;

import com.bst.pmgr.entities.Document;
import com.bst.pmgr.entities.Work;
import com.bst.pmgr.entities.components.DocumentListener;
import com.bst.pmgr.entities.repositories.DocumentRepository;
import com.bst.pmgr.entities.schema.SchemaService;
import com.bst.utility.components.RepositoryAspect;
import com.bst.utility.testlib.SnapshotListener;

@ExtendWith(SpringExtension.class)
@TestExecutionListeners(listeners = SnapshotListener.class, mergeMode = MergeMode.MERGE_WITH_DEFAULTS)
@DataJpaTest
@ContextConfiguration(classes = { TestConfiguration.class })
@TestPropertySource("classpath:global-test.properties")
public class DocumentListenerTest {

	@TestConfiguration
	@EntityScan("com.bst.pmgr.entities")
	@EnableJpaRepositories("com.bst.pmgr.entities.repositories")
	@EnableAspectJAutoProxy
	static class DocumentListenerTestConfiguration {

		@Bean
		public DocumentListener getDocumentListener() {
			return new DocumentListener();
		}

		@Bean
		public RepositoryAspect getRepositoryAspect() {
			return new RepositoryAspect();
		}

		@Bean
		public SchemaService getSchemaService() {
			return new SchemaService();
		}

	}

	@FunctionalInterface
	public interface TransactionWrappedFunction {
		public void wrap() throws Exception;
	}

	@Autowired
	private DocumentRepository documentRepository;

	@SuppressWarnings("unused")
	@Autowired
	private DocumentListener documentListener;

	@Test
	public void testDocumentListenerInitializeDocument() throws Exception {

		this.transactionWrapper(() -> {
			final Work work = new Work();
			work.setName("test work");

			final Document document = new Document();
			document.setMetaName("projectStaffAssignments");
			work.addDocument(document);

			this.documentRepository.save(document);
		});

		this.transactionWrapper(() -> {
			Document document = this.documentRepository.findAll().iterator().next();
			SnapshotListener.expect(document).toMatchSnapshot();

			document.removeTool(document.getTools().get(0));
			SnapshotListener.expect(document).toMatchSnapshot();

			document.setName("new name");

			document = this.documentRepository.save(document);
		});

		final Document document = this.documentRepository.findAll().iterator().next();
		SnapshotListener.expect(document).toMatchSnapshot();
	}

	@Transactional(propagation = Propagation.REQUIRES_NEW)
	private void transactionWrapper(final TransactionWrappedFunction fn) throws Exception {
		fn.wrap();
	}

}
