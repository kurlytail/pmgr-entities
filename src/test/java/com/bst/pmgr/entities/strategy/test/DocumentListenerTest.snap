{
  "testDocumentListenerInitializeDocument" : [ {
    "id" : 1,
    "metaName" : "projectStaffAssignments",
    "name" : "constructive_carrot",
    "work" : {
      "id" : 2,
      "name" : "test work",
      "activities" : [ ],
      "documents" : [ 1 ],
      "tools" : [ {
        "id" : 3,
        "name" : "flexible_designer",
        "metaName" : "acquisition",
        "work" : 2,
        "document" : 1,
        "executionPercent" : 0
      }, {
        "id" : 4,
        "name" : "linear_fund",
        "metaName" : "preassignment",
        "work" : 2,
        "document" : 1,
        "executionPercent" : 0
      }, {
        "id" : 5,
        "name" : "popular_leaf",
        "metaName" : "virtualTeams",
        "work" : 2,
        "document" : 1,
        "executionPercent" : 0
      }, {
        "id" : 6,
        "name" : "steady_parent",
        "metaName" : "negotiation",
        "work" : 2,
        "document" : 1,
        "executionPercent" : 0
      } ],
      "processes" : [ ],
      "processGroups" : [ ]
    },
    "tools" : [ 3, 4, 5, 6 ],
    "sections" : [ ]
  }, {
    "id" : 1,
    "metaName" : "projectStaffAssignments",
    "name" : "constructive_carrot",
    "work" : {
      "id" : 2,
      "name" : "test work",
      "activities" : [ ],
      "documents" : [ 1 ],
      "tools" : [ {
        "id" : 4,
        "name" : "linear_fund",
        "metaName" : "preassignment",
        "work" : 2,
        "document" : 1,
        "executionPercent" : 0
      }, {
        "id" : 5,
        "name" : "popular_leaf",
        "metaName" : "virtualTeams",
        "work" : 2,
        "document" : 1,
        "executionPercent" : 0
      }, {
        "id" : 6,
        "name" : "steady_parent",
        "metaName" : "negotiation",
        "work" : 2,
        "document" : 1,
        "executionPercent" : 0
      } ],
      "processes" : [ ],
      "processGroups" : [ ]
    },
    "tools" : [ 4, 5, 6 ],
    "sections" : [ ]
  }, {
    "id" : 1,
    "metaName" : "projectStaffAssignments",
    "name" : "new name",
    "work" : {
      "id" : 2,
      "name" : "test work",
      "activities" : [ ],
      "documents" : [ 1 ],
      "tools" : [ {
        "id" : 4,
        "name" : "linear_fund",
        "metaName" : "preassignment",
        "work" : 2,
        "document" : 1,
        "executionPercent" : 0
      }, {
        "id" : 5,
        "name" : "popular_leaf",
        "metaName" : "virtualTeams",
        "work" : 2,
        "document" : 1,
        "executionPercent" : 0
      }, {
        "id" : 6,
        "name" : "steady_parent",
        "metaName" : "negotiation",
        "work" : 2,
        "document" : 1,
        "executionPercent" : 0
      }, {
        "id" : 7,
        "name" : "abstract_rest",
        "metaName" : "acquisition",
        "work" : 2,
        "document" : 1,
        "executionPercent" : 0
      } ],
      "processes" : [ ],
      "processGroups" : [ ]
    },
    "tools" : [ 4, 5, 6, 7 ],
    "sections" : [ ]
  } ]
}